### PR TITLE
Pick Depth Buffer

### PR DESCRIPTION
**Pull request by bagnell**
NOTE: This is a PR into the globe-depth branch.

Adds `Scene.pickDepth` that takes window coordinates and returns the `Cartesian3` world position. Also uses this for mouse interaction when available.

### CODE REVIEW

**Comment by pjcozzi** [comment]
For a 1920x1080 display, each RGBA8 texture for depth will be 8,294,400 bytes.  Lets run some memory numbers between this, `globe-depth`, and `master` as you suggested offline.

File: Source/Scene/PickDepth.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        destroyFramebuffers(pickDepth);
+
+        createTextures(pickDepth, context, width, height);
+
+        pickDepth.framebuffer = context.createFramebuffer({
+            colorTextures : [pickDepth._depthTexture],
+            destroyAttachments : false
+        });
+
+        if (pickDepth.framebuffer.status !== WebGLRenderingContext.FRAMEBUFFER_COMPLETE) {
```
Comment (pjcozzi) [inline Source/Scene/PickDepth.js]: Can we add a comment about why this could fail?

File: Source/Scene/PickDepth.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    function updateCopyCommands(pickDepth, context, depthTexture) {
+        if (!defined(pickDepth._copyDepthCommand)) {
+            var fs = 'uniform sampler2D u_texture;\n' +
+                     'varying vec2 v_textureCoordinates;\n' +
+                     'void main()\n' +
+                     '{\n' +
+                     '    float depth = texture2D(u_texture, v_textureCoordinates).r;\n' +
+                     '    float temp = depth * 255.0;\n' +
```
Comment (pjcozzi) [inline Source/Scene/PickDepth.js]: Think this is generic enough for a `czm_` function, e.g., for a shadow mapping fallback?

Comment (pjcozzi) [inline Source/Scene/PickDepth.js]: CC @fstoner

File: Source/Scene/Scene.js [hunk 3]
```diff
@@ -1210,6 +1215,15 @@ define([
         return globeDepth;
     }
 
+    function getPickDepth(scene, context, index) {
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: We don't need to pass `context`, right?  Just use `scene.context`.

File: Source/Scene/Scene.js [hunk 4]
```diff
@@ -1397,6 +1411,10 @@ define([
             commands = frustumCommands.commands[Pass.TRANSLUCENT];
             commands.length = frustumCommands.indices[Pass.TRANSLUCENT];
             executeTranslucentCommands(scene, executeCommand, passState, commands);
+
+            var pickDepth = getPickDepth(scene, context, index);
+            pickDepth.update(context, globeDepth.framebuffer.depthStencilTexture);
+            pickDepth.executeCopyDepth(context, passState);
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: We didn't implement the MRT optimization right?  Thinking about it more...MRT _could_ be slower, for example, if the scene has a high depth complexity it would result in a lot more memory bandwidth.  I don't think this warrants a roadmap issue, perhaps add `PERFORMANCE_IDEA` comment in case we run into this extra pass when profiling?

File: Source/Scene/Scene.js [hunk 5]
```diff
@@ -1698,6 +1716,95 @@ define([
...
+    var scratchWorldCoords = new Cartesian4();
+
+    /**
+     * Returns the cartesian position reconstructed from the depth buffer and window position.
+     *
+     * @param {Cartesian2} windowPosition Window coordinates to perform picking on.
+     * @param {Cartesian3} [result] The object on which to restore the result.
+     * @returns {Cartesian3} The cartesian position.
+     *
+     * @exception {DeveloperError} windowPosition is undefined.
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: This doesn't need to be documented.  It is implicit above.

File: Source/Scene/Scene.js [hunk 6]
```diff
@@ -1698,6 +1716,95 @@ define([
...
+     * @exception {DeveloperError} Picking from the depth buffer is not supported.
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Can we tell the user the function to call to check?

File: Source/Scene/Scene.js [hunk 7]
```diff
@@ -1698,6 +1716,95 @@ define([
...
+        var camera = this._camera;
+
+        // Create a working frustum from the original camera frustum.
+        var frustum;
+        if (defined(camera.frustum.fov)) {
+            frustum = camera.frustum.clone(scratchPerspectiveFrustum);
+        } else if (defined(camera.frustum.infiniteProjectionMatrix)){
+            frustum = camera.frustum.clone(scratchPerspectiveOffCenterFrustum);
+        } else {
+            throw new DeveloperError('An orthographic projection matrix is not invertible.');
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: This means we are in 2D mode, right?  Can we say that?

File: Source/Scene/Scene.js [hunk 8]
```diff
@@ -1698,6 +1716,95 @@ define([
...
+        }
+
+        var renderedFrustum = this._frustumCommandsList[index];
+        frustum.near = renderedFrustum.near * 0.99;
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: This assumes opaque geometry, right?  Probably not a bad assumption, but can we make the `0.99` a file-level constant, e.g., `OPAQUE_FRUSTUM_NEAR_OFFSET` or whatever?

File: Source/Scene/Scene.js [hunk 9]
```diff
@@ -1698,6 +1716,95 @@ define([
...
+        frustum.far = renderedFrustum.far;
+        uniformState.updateFrustum(frustum);
+
+        var viewport = uniformState.viewport;
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Make this a separate function perhaps in `Transforms` since it goes well with [pointToWindowCoordinates](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Transforms.js#L740).  It can be `@private` until it is needed.

File: Source/Scene/ScreenSpaceCameraController.js [hunk 10]
```diff
@@ -595,8 +595,18 @@ define([
         var origin = Cartesian3.clone(Cartesian3.ZERO, translateCVOrigin);
         var normal = Cartesian3.UNIT_X;
 
-        if (defined(controller._globe) && camera.position.z < controller.minimumPickingTerrainHeight) {
-            var intersection = controller._globe.pick(startRay, scene, translateCVStartPos);
+        var globe = controller._globe;
+        if (defined(globe) && camera.position.z < controller.minimumPickingTerrainHeight) {
```
Comment (pjcozzi) [inline Source/Scene/ScreenSpaceCameraController.js]: Similar code is duplicated a few times in this file.  Can we clean it up?

**Comment by pjcozzi** [comment]
This needs tests.

File: Source/Scene/PickDepth.js (hunk 1 above)
Comment (bagnell) [inline Source/Scene/PickDepth.js]: This was here because the color texture was float, but later I packed the depth into 3 unsigned bytes. According to the spec, a single RGBA/unsigned byte color attachment must be complete. I'll remove the check.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look.

File: Source/Scene/Scene.js [hunk 11]
```diff
@@ -1729,6 +1763,79 @@ define([
...
+     * @param {Cartesian3} [result] The object on which to restore the result.
+     * @returns {Cartesian3} The cartesian position.
+     *
+     * @exception {DeveloperError} Picking from the depth buffer is not supported. Check Context.depthTexture for support.
+     * @exception {DeveloperError} 2D is not supported. An orthographic projection matrix is not invertible.
+     */
+    Scene.prototype.pickDepth = function(windowPosition, result) {
+        //>>includeStart('debug', pragmas.debug);
+        if(!defined(windowPosition)) {
+            throw new DeveloperError('windowPosition is undefined.');
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Do we have a test for this?

File: CHANGES.md [hunk 12]
```diff
@@ -15,6 +15,7 @@ Change Log
 * Fixed documentation for 'Context.drawingBufferHeight' and 'Context.drawingBufferWidth'.
 * Added debug option to 'Scene' to show the depth buffer information for a specified view frustum slice and exposed capability in 'CesiumInspector' widget.
 * Removed `Scene.fxaaOrderIndependentTranslucency`. Use `Scene.fxaa` which is now `true` by default.
+* Added `Scene.pickDepth` to reconstruct the cartesian world position from window coordinates.
```
Comment (pjcozzi) [inline CHANGES.md]: Can we mention that users can now pan/zoom/rotate around 3D models and other opaque geometry?

**Comment by bagnell** [comment]
@pjcozzi This is ready.

**Comment by pjcozzi** [comment]
@bagnell what's left here?  Just the `undefined` values being returned in the mouse-over example?

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look. If you zoom out far enough, you get some strange height values. I think most of the time it is returning a value for the depth plane.

**Comment by pjcozzi** [comment]
Is this the strange value?

image

I believe this is the depth plane, but shouldn't we be able to return a "correct" result from the users perspective?  Why should the depth plane win in the above screenshot?

**Comment by pjcozzi** [comment]
Do we need new unit tests for the corner cases?

**Comment by pjcozzi** [comment]
Why did we switch away from the Cesium label? Was there a Cesium bug? Users are already looking at the mouse so its not clear to me why we would want to display the data elsewhere.

**Comment by bagnell** [comment]
>Is this the strange value?

Yes

>I believe this is the depth plane, but shouldn't we be able to return a "correct" result from the users perspective? Why should the depth plane win in the above screenshot?

I think that is the first depth value we get. Otherwise we would need to read pixels from all of the frustums and return the closest.

**Comment by bagnell** [comment]
>Why did we switch away from the Cesium label?

The depth test and eye offset were causing the label to disappear.

**Comment by pjcozzi** [comment]
> Otherwise we would need to read pixels from all of the frustums and return the closest.

Sounds like that is what we are going to have to do then unless we, for example, add a pass that stencils out the depth plane.  This will only be a problem until we have vector data on terrain, then we will not need the depth plane.

**Comment by pjcozzi** [comment]
> The depth test and eye offset were causing the label to disappear.

Because we were moving it forward out of view?  Do you think we could just use an eye offset on the order of the model's bounding sphere?

**Comment by bagnell** [comment]
>Because we were moving it forward out of view? Do you think we could just use an eye offset on the order of the model's bounding sphere?

That works OK. There are still some multi-frustum issues showing/disappearing when the depth plane is picked. 

Getting the right bounding sphere for a model is a huge hack. I can't get the bounding sphere from the entity so I need to use `Scene/Model` which isn't a big deal just much more code. To get the bounding sphere I need to go through the private `_nodeCommands` and get the biggest radius. the `boundingSphere` property only returns the radius of the un-scaled model (without `minimumPixelSize`).

**Comment by pjcozzi** [comment]
>  the boundingSphere property only returns the radius of the un-scaled model (without minimumPixelSize).

Can you add something that returns the scaled bounding sphere or radius?  A pull request to master would be fine.

**Comment by mramato** [comment]
> I can't get the bounding sphere from the entity

Is this in example code?  If yes, then you are correct, @pjcozzi asked not to expose it until we had a reason to do so but there is a way to get it.  If you're not in example code, then you can get the BoundingSphere for an entity, but you need a handle the the viewer (more specifically, the `DataSourceDsiplay` instance that lives in the viewer).

**Comment by pjcozzi** [comment]
> Can you add something that returns the scaled bounding sphere or radius? A pull request to master would be fine.

We could also keep our eye on the prize here and just turn off the minimum pixel size so we can use the regular bounding sphere.  This is probably the better option unless you've already made progress here.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another review.

File: Source/Scene/Scene.js [hunk 13]
```diff
@@ -1748,6 +1787,88 @@ define([
...
+     * Returns the cartesian position reconstructed from the depth buffer and window position.
+     *
+     * @param {Cartesian2} windowPosition Window coordinates to perform picking on.
+     * @param {Cartesian3} [result] The object on which to restore the result.
+     * @returns {Cartesian3} The cartesian position.
+     *
+     * @exception {DeveloperError} Picking from the depth buffer is not supported. Check Context.depthTexture for support.
+     * @exception {DeveloperError} 2D is not supported. An orthographic projection matrix is not invertible.
+     */
+    Scene.prototype.pickDepth = function(windowPosition, result) {
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: I should have thought of this sooner, but don't you think this should be `pickPosition` since `pickDepth` is an implementation detail?

File: Source/Scene/ScreenSpaceCameraController.js [hunk 14]
```diff
@@ -566,6 +566,37 @@ define([
...
+        var scene = controller._scene;
+        var globe = controller._globe;
+        var camera = scene.camera;
+
+        if (!defined(globe)) {
+            return undefined;
+        }
+
+        var depthIntersection;
+        if (defined(scene._globeDepth)) {
```
Comment (pjcozzi) [inline Source/Scene/ScreenSpaceCameraController.js]: Can we have a public helper so users can also see if this is supported?  The Sandcastle example should also check it.

Comment (bagnell) [inline Source/Scene/ScreenSpaceCameraController.js]: Can we just check `Context.depthTexture` instead? It will be defined if it is supported.

Comment (pjcozzi) [inline Source/Scene/ScreenSpaceCameraController.js]: `Context` is not public, we need to provide access from the `Scene.

**Comment by pjcozzi** [comment]
I get six test failures.

**Comment by bagnell** [comment]
>I get six test failures.

I missed a rename. They're fixed now.

**Comment by bagnell** [comment]
@pjcozzi This is ready unless you have any other comments.

**Comment by pjcozzi** [comment]
All looks good but remind me why the label still blinks sometimes when zooming in close and mousing over the truck?

**Comment by bagnell** [comment]
I'm not sure. I haven't seen the blinking. I have seen it disappear when zoomed all the way in. That's the eye offset. I set it to the diameter of the bounding sphere so it disappears because it's behind the camera.

**Comment by pjcozzi** [comment]
I noticed the disappearing, that's why I submitted #2742.  The blinking is different.  It is most occurs on the sides of the truck, not the roof.

**Comment by pjcozzi** [comment]
All good.
